File: Windows/FireflyDevice/FDError.cpp
//
//  FDError.cpp
//  FireflyDevice
//
//  Created by Denis Bohm on 3/25/14.
//  Copyright (c) 2014 Firefly Design LLC / Denis Bohm. All rights reserved.
//

#include "FDError.h"
#include "FDString.h"

#include <memory>

namespace FireflyDesign {

	std::shared_ptr<FDError> FDError::error(std::string domain, int code, std::map<std::string, std::string> userInfo) {
		std::shared_ptr<FDError> error = std::make_shared<FDError>();
		error->domain = domain;
		error->code = code;
		error->userInfo = userInfo;
		return error;
	}

	std::shared_ptr<FDError> FDError::error(std::string domain, int code, std::string description) {
		std::map<std::string, std::string> userInfo;
		userInfo[FDLocalizedDescriptionKey] = description;
		return FDError::error(domain, code, userInfo);
	}

	std::string FDError::description() {
		std::string s = userInfo[FDLocalizedDescriptionKey];
		return FDString::format("%s %u %s", domain.c_str(), code, s.c_str());
	}

}

File: Android/Libraries/fireflydevice/src/main/java/com/fireflydesign/fireflydevice/FDError.java
//
//  FDError.cpp
//  FireflyDevice
//
//  Created by Denis Bohm on 3/25/14.
//  Copyright (c) 2014 Firefly Design LLC / Denis Bohm. All rights reserved.
//

package com.fireflydesign.fireflydevice;

import java.util.HashMap;
import java.util.Map;

public class FDError {

    public static final String FDLocalizedDescriptionKey = "description";
    public static final String FDLocalizedRecoveryOptionsErrorKey = "recoveryOptionsError";

    public static FDError error(String domain, int code, Map<String, String> userInfo) {
		FDError error = new FDError();
		error.domain = domain;
		error.code = code;
		error.userInfo = userInfo;
		return error;
	}

    public static FDError error(String domain, int code, String description) {
		Map<String, String> userInfo = new HashMap<String, String>();
		userInfo.put(FDLocalizedDescriptionKey, description);
		return FDError.error(domain, code, userInfo);
	}

	public String description() {
		String s = userInfo.get(FDLocalizedDescriptionKey);
		return FDString.format("%s %u %s", domain, code, s);
	}

    public String domain;
    public int code;
    public Map<String, String> userInfo;

}
